oak_crypto/src/noise_nk/noise.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This was copied from Chromium's third_party/cloud_authenticator, which has
+//! compatible copyright and ownership (Apache 2.0, Google).
+
+use alloc::vec::Vec;
+
+use crate::noise_nk::{error::Error, rustcrypto};
+
+const SYMMETRIC_KEY_SIZE: usize = 32;
+pub const NONCE_LEN: usize = 12;
+
+#[derive(Debug, PartialEq)]
+pub enum HandshakeType {
+    Nk, // https://noiseexplorer.com/patterns/NK/
+}
+
+// Helper to generate 2 keys.
+fn hkdf2(
+    ck: &[u8; SYMMETRIC_KEY_SIZE],
+    ikm: &[u8],
+) -> ([u8; SYMMETRIC_KEY_SIZE], [u8; SYMMETRIC_KEY_SIZE]) {
+    let mut output = [0; rustcrypto::SHA256_OUTPUT_LEN * 2];
+    // unwrap: only fails if the output size is too large, but the output
+    // size is small and fixed here.
+    rustcrypto::hkdf_sha256(ikm, ck, &[], &mut output).unwrap();
+    let key1: [u8; SYMMETRIC_KEY_SIZE] = output[..SYMMETRIC_KEY_SIZE].try_into().unwrap();
+    let key2: [u8; SYMMETRIC_KEY_SIZE] = output[SYMMETRIC_KEY_SIZE..].try_into().unwrap();
+    (key1, key2)
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Noise {
+    chaining_key: [u8; SYMMETRIC_KEY_SIZE],
+    h: [u8; SYMMETRIC_KEY_SIZE],
+    symmetric_key: [u8; SYMMETRIC_KEY_SIZE],
+    symmetric_nonce: u32,
+}
+
+impl Noise {
+    pub fn new(handshake_type: HandshakeType) -> Self {
+        assert_eq!(handshake_type, HandshakeType::Nk);
+        let mut chaining_key_in = [0; SYMMETRIC_KEY_SIZE];
+        let nk_protocol_name = b"Noise_NK_P256_AESGCM_SHA256";
+        chaining_key_in[..nk_protocol_name.len()].copy_from_slice(nk_protocol_name);
+        Noise {
+            chaining_key: chaining_key_in,
+            h: chaining_key_in,
+            symmetric_key: [0; SYMMETRIC_KEY_SIZE],
+            symmetric_nonce: 0,
+        }
+    }
+
+    pub fn mix_hash(&mut self, in_data: &[u8]) {
+        // See https://www.noiseprotocol.org/noise.html#the-symmetricstate-object
+        self.h = rustcrypto::sha256_two_part(&self.h, in_data);
+    }
+
+    pub fn mix_key(&mut self, ikm: &[u8]) {
+        // See https://www.noiseprotocol.org/noise.html#the-symmetricstate-object
+        let derived_keys = hkdf2(&self.chaining_key, ikm);
+        self.chaining_key.copy_from_slice(&derived_keys.0);
+        self.initialize_key(&derived_keys.1);
+    }
+
+    #[cfg(test)]
+    pub fn mix_key_and_hash(&mut self, ikm: &[u8]) {
```

Comment:
Why is this function test-only?

Reply:
It is meant to be used outside of tests, but so far it is only used in tests.  Clippy complains about an unused function if we remove that.